R/utils-callbacks.R

```diff
@@ -0,0 +1,164 @@
+#' invoked sequentially with a single call to `$invoke()` with an object or data
+#' to pass to the callback functions. Callbacks must take at least one argument
+#' are invoked in reverse order of their registration.
+#'
+#' @noRd
+CallbackManager <- R6Class(
+  "CallbackManager",
+
+  public = list(
+    initialize = function(args = NULL) {
+      private$args <- args
+    },
+
+    #' @description Add a callback function.
+    #' @param callback A function to be called.
+    #' @return A function that can be called to remove the callback.
+    add = function(callback, call = caller_env()) {
+      check_function2(callback, args = private$args, call = call)
+
+      id <- private$next_id()
+      private$callbacks[[id]] <- callback
+
+      fn_remove <- function() {
+        private$callbacks[[id]] <- NULL
+        invisible(NULL)
+      }
+      invisible(fn_remove)
+    },
+
+    #' @description
+    #' Invoke all registered callbacks with the provided arguments. Callbacks
+    #' are invoked in reverse order of registration (last-in first-evaluated).
+    #'
+    #' @param ... Arguments to pass to the callbacks.
+    #' @returns Nothing, callbacks are invoked for side effects).
+    invoke = function(...) {
+      if (length(private$callbacks) == 0) {
+        return(invisible(NULL))
+      }
+
+      # Invoke callbacks in reverse insertion order
+      for (id in rev(as.integer(names(private$callbacks)))) {
+        res <- exec(private$callbacks[[as.character(id)]], ...)
+        if (promises::is.promise(res)) {
+          cli::cli_abort(c(
+            "Can't use async callbacks with `$chat()` or `$stream()`.",
+            i = "Async callbacks are supported, but you must use `$chat_async()` or `$stream_async()`."
+          ))
+        }
+      }
+
+      invisible(NULL)
+    },
+
+    #' @description
+    #' Invoke all registered callbacks asynchronously with the provided
+    #' arguments. As with `$invoke()`, callbacks are invoked in reverse order of
+    #' registration (last-in first-evaluated).
+    #'
+    #' @param ... Arguments to pass to the callbacks.
+    #' @returns Nothing, callbacks are invoked for side effects).
+    invoke_async = async_method(function(self, private, ...) {
+      if (length(private$callbacks) == 0) {
+        return(invisible(NULL))
+      }
+
+      # Invoke callbacks in reverse insertion order
+      for (id in rev(as.integer(names(private$callbacks)))) {
+        coro::await(exec(private$callbacks[[as.character(id)]], ...))
+      }
+
+      invisible(NULL)
+    }),
+
+    #' @description Get the number of registered callbacks.
+    #' @return Integer count of callbacks.
+    count = function() {
+      length(private$callbacks)
+    },
+
+    #' @description Clear all registered callbacks.
+    clear = function() {
+      private$callbacks <- list()
+      invisible(NULL)
+    },
+
+    #' @describeIn Get callback list
+    peek = function() {
+      private$callbacks
```

Comment:
FWIW I'd expect `peek` to reveal just the last element, as it's typically paired with `pop()`, which removes and returns the last element.